src/lib/pubkey/hss_lms/hss_lms.cpp

```diff
@@ -0,0 +1,196 @@
+}
+
+HSS_LMS_PrivateKey::HSS_LMS_PrivateKey(std::span<const uint8_t> private_key) {
+   m_private = HSS_LMS_PrivateKeyInternal::from_bytes_or_throw(private_key);
+   m_public = std::make_shared<HSS_LMS_PublicKeyInternal>(HSS_LMS_PublicKeyInternal::create(*m_private));
+}
+
+HSS_LMS_PrivateKey::HSS_LMS_PrivateKey(RandomNumberGenerator& rng, std::string_view algo_params) {
+   HSS_LMS_Params hss_params(algo_params);
+   m_private = std::make_shared<HSS_LMS_PrivateKeyInternal>(hss_params, rng);
+   m_public = std::make_shared<HSS_LMS_PublicKeyInternal>(HSS_LMS_PublicKeyInternal::create(*m_private));
+}
+
+HSS_LMS_PrivateKey::HSS_LMS_PrivateKey(std::shared_ptr<HSS_LMS_PrivateKeyInternal> sk) : m_private(std::move(sk)) {
+   m_public = std::make_shared<HSS_LMS_PublicKeyInternal>(HSS_LMS_PublicKeyInternal::create(*m_private));
+}
+
+HSS_LMS_PrivateKey::~HSS_LMS_PrivateKey() = default;
+
+secure_vector<uint8_t> HSS_LMS_PrivateKey::private_key_bits() const {
+   return m_private->to_bytes();
+}
+
+secure_vector<uint8_t> HSS_LMS_PrivateKey::raw_private_key_bits() const {
+   return private_key_bits();
+}
+
+std::unique_ptr<Public_Key> HSS_LMS_PrivateKey::public_key() const {
+   return std::make_unique<HSS_LMS_PublicKey>(*this);
+}
+
+// We use a separate algorithm identifier for the private key since we use a Botan scoped OID for it.
+// This is necessary since the private key format is implementation specific, since it is not defined
+// in RFC 8554.
+AlgorithmIdentifier HSS_LMS_PrivateKey::pkcs8_algorithm_identifier() const {
+   return AlgorithmIdentifier(OID::from_string("HSS-LMS-Private-Key"), AlgorithmIdentifier::USE_EMPTY_PARAM);
+}
+
+std::optional<uint64_t> HSS_LMS_PrivateKey::remaining_operations() const {
+   return (m_private->hss_params().max_sig_count() - m_private->get_idx()).get();
+}
+
+std::unique_ptr<Private_Key> HSS_LMS_PrivateKey::generate_another(RandomNumberGenerator& rng) const {
+   // Cannot use std::make_unique because the utilized constructor is private.
+   return std::unique_ptr<HSS_LMS_PrivateKey>(
+      new HSS_LMS_PrivateKey(std::make_shared<HSS_LMS_PrivateKeyInternal>(m_private->hss_params(), rng)));
+}
+
+class HSS_LMS_Signature_Operation final : public PK_Ops::Signature {
+   public:
+      HSS_LMS_Signature_Operation(std::shared_ptr<HSS_LMS_PrivateKeyInternal> private_key,
+                                  std::shared_ptr<HSS_LMS_PublicKeyInternal> public_key) :
+            m_private(std::move(private_key)), m_public(std::move(public_key)) {}
+
+      void update(const uint8_t msg[], size_t msg_len) override {
+         m_msg_buffer.insert(m_msg_buffer.end(), msg, msg + msg_len);
+      }
+
+      secure_vector<uint8_t> sign(RandomNumberGenerator&) override {
+         std::vector<uint8_t> message_to_sign = std::exchange(m_msg_buffer, {});
+         return m_private->sign(message_to_sign);
+      }
+
+      size_t signature_length() const override { return HSS_Signature::size(m_private->hss_params()); }
```

Comment:
This length could be computed and cached once, either here or in the private key itself.